Add Favorite Servers!

staticfox opened the pull request:
Added favorites, thoughts?

Checklist of stuff I still have to do
- [x] Implement favorite base
- [x] Add  icon
- [x] Add right click button for favoriting

---

Fixes #1477

staticfox committed:
Add base work for favorites

un1versal committed:
KviIconManager: add new icon serverfavorite

and cleanup unused and replaced icon

un1versal committed:
assign serverfavorite icon to function

staticfox committed:
favorites: Add ServerFavorite to the browser

staticfox committed:
server: Statically cast flags

un1versal commented on src/modules/options/OptionsWidget_servers.cpp:
```diff
@@ -1817,8 +1842,8 @@ void OptionsWidget_servers::commit()
 
 void OptionsWidget_servers::customContextMenuRequested(const QPoint &pnt)
```
This is a cosmetic thing I noticed when sending PR,
Do you think its worth adding a indenting fix for this function below so all entries are not all over? 

like line 1849, 1860 1862, 1863 etc.

un1versal commented:
+1 and I have a Q, will the favorited server be marked on the recent server list also?

You can also drop the merge commit message by doing a interactive rebase.

staticfox committed:
favorites: Fix indentation

staticfox commented:
fixed

un1versal commented:
OK some suggestions/feedback as discussed on IRC which seem to make sense. I dont mean to be pedantic just somewhat thorough. :smile: 

1) When a Server is already a favorite the label should say **Unfavorite Server** both on **right click menu label** and **toolbox tooltip**
![capture](https://cloud.githubusercontent.com/assets/3521959/10930020/6163ed72-82b2-11e5-83d4-312682dbd8ee.PNG)

2) When **Only display favorite server** is enabled, and you **Unfavorite a Server** the list should be refreshed automatically and the Unfavorited server removed form list.
![capture1](https://cloud.githubusercontent.com/assets/3521959/10929863/a2981202-82b0-11e5-91f6-36f6436aeb24.PNG)

3) **Recent Server Lists** If the recent server was a favorite the entry on these lists should also reflect this IMO. Theres 3 of these instances

**Configure servers dialogs** Recent Servers List
![capture](https://cloud.githubusercontent.com/assets/3521959/10930404/a931fb5a-82b5-11e5-928b-1520f54db3c1.PNG)

**Configure KVIrc servers dialogs** Recent Servers List
![capture](https://cloud.githubusercontent.com/assets/3521959/10930598/a3737be2-82b7-11e5-8619-73fcce03a6dc.PNG)

Main KVIrc Toolbar, Recent Server Lists
![capture](https://cloud.githubusercontent.com/assets/3521959/10930752/0c370e9a-82b9-11e5-8011-f968b8273bb8.PNG)

This next one is somewhat related, I think it would be nice to address it while messing around with these, but could be done later.

4)The toolbox entries would look better and help usability if they would be grouped by similar akin functions and reordered

**Current**
![capture](https://cloud.githubusercontent.com/assets/3521959/10930285/b3673e10-82b4-11e5-8013-b6bc8d2b0e72.PNG)

**Mockup of suggestion** 
![capture1](https://cloud.githubusercontent.com/assets/3521959/10930293/c1f5b5ec-82b4-11e5-872c-bd4fa918da44.PNG)

![capture](https://cloud.githubusercontent.com/assets/3521959/10930539/082c2184-82b7-11e5-999f-6af71b5687c7.PNG)

staticfox committed:
favorites: Hide server when unfavorited

Have the menu option labeled 'Unfavorite Server' when it is
currently favorited.

staticfox commented:
1) Done, however it's not really possible to do on the right side since it's a static button.
2) Done
3) Still looking in to.. it's weird. :|
4) I'd like to get others feedback on that first

staticfox committed:
Small cleanups

staticfox commented:
Alright so after looking this over, I think I'm at a good point for this to go in. re: @un1versal 

> 4)The toolbox entries would look better and help usability if they would be grouped by similar akin functions and reordered

IMO It seems pretty in order as is. You have adding and deleting servers at the top group, copy/paste in the middle, and favoriting at the bottom.

Aside from that, anyone have anything they would like to add?

un1versal commented:
Not really just that the favorite server should be 3rd entry and only have one separator so like

```
  add network 
  add server 
  favorite server
  remove
  copy
  paste
  -------
  Import servers
```

That feels more natural, then what is atm; however we can do that in separate PR
Lets give this one more spin and merge it.

staticfox commented:
The above can be debated about in a ticket, but for now, I'm just going to pull this in.
